File: 2024/01.java
/*
 * Advent of Code 2024
 * Day 1: Historian Hysteria
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Hashtable;
import java.util.Scanner;


class Day01 {
    private ArrayList<Integer> listA = new ArrayList<Integer>();
    private ArrayList<Integer> listB = new ArrayList<Integer>();

    private void getInput() {
        try {
            Scanner fileReader = new Scanner(new File("input.txt"));
            while (fileReader.hasNextLine()) {
                String line = fileReader.nextLine();
                listA.add(Integer.parseInt(line.split("   ")[0]));
                listB.add(Integer.parseInt(line.split("   ")[1]));
            }
            fileReader.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            return;
        }
    }

    private void partOne() {
        int distance = 0;
        listA.sort(Comparator.naturalOrder());
        listB.sort(Comparator.naturalOrder());
        for (int i = 0; i < listA.size(); i++) {
            distance += Math.abs(listA.get(i) - listB.get(i));
        }
        System.out.println(String.format("Part one: %d", distance));
    }

    private void partTwo() {
        int distance = 0;
        Hashtable<Integer, Integer> occurences = new Hashtable<>();
        for (int num : listB) {
            if (occurences.containsKey(num)) {
                occurences.put(num, occurences.get(num) + 1);
            } else {
                occurences.put(num, 1);
            }
        }
        for (int num : listA) {
            distance += occurences.containsKey(num) ? occurences.get(num) * num : 0;
        }
        System.out.println(String.format("Part two: %d", distance));
    }

    public static void main(String[] args) {
        Day01 solution = new Day01();
        solution.getInput();
        solution.partOne();
        solution.partTwo();
    }
}

File: 2024/01.py
def parse_input():
    list_a = []
    list_b = []
    with open('input.txt', 'r') as _input:
        for line in _input: 
            line = line.strip().split('   ')
            list_a.append(int(line[0]))
            list_b.append(int(line[1]))
    return list_a, list_b

def part_one(list_a, list_b):
    total_distance = 0
    list_a.sort()
    list_b.sort()
    for i in range(len(list_a)):
        total_distance += abs(list_a[i] - list_b[i])
    print(total_distance)


def part_two(list_a, list_b):
    total_distance = 0
    occurences = {}
    for num in list_b:
        occurences[num] = occurences[num] + 1 if num in occurences else 1
    for num in list_a:
        total_distance += num * occurences[num] if num in occurences else 0
    print(total_distance)


if __name__ == '__main__':
    list_a, list_b = parse_input()
    part_one(list_a, list_b)
    part_two(list_a, list_b)
